main/src/ca/uwaterloo/flix/language/ast/Type.scala

```diff
@@ -945,4 +946,72 @@ object Type {
     }
   }
 
+  /**
+    * Checks that `tpe1` is a subtype of `tpe2`.
+    *
+    * `tpe1` is a subtype of `tpe2` if:
+    *
+    * (a) `tpe1` has the exact same flix type as `tpe2`
+    *
+    * (b) both types are java types and `tpe1` is a subtype of `tpe2`
+    *
+    * (c) both types are functions and `tpe1` is a subtype of `tpe2`
+    *
+    * AND
+    *
+    * the purity of the expression is being cast from `pure` -> `ef` -> `impure`.
+    *
+    * OR
+    *
+    * the effect set of the expression is a subset of the effect set being cast to.
+    *
+    */
+  def isRegularSubType(tpe1: Type, tpe2: Type, renv: RigidityEnv)(implicit flix: Flix): Boolean = (tpe1.baseType, tpe2.baseType) match {
+    case (Type.Cst(TypeConstructor.Native(left), _), Type.Cst(TypeConstructor.Native(right), _)) =>
+      right.isAssignableFrom(left)
+
+    case (Type.Cst(TypeConstructor.Arrow(n1), _), Type.Cst(TypeConstructor.Arrow(n2), _)) if n1 == n2 =>
+      // check that parameters are supertypes
+      val args1 = tpe1.arrowArgTypes
+      val args2 = tpe2.arrowArgTypes
+      val superTypeArgs = args1.zip(args2).forall {
+        case (t1, t2) =>
+          Type.isRegularSubType(t2, t1, renv)
+      }
+
+      // check that result is a subtype
+      val expectedResTpe = tpe1.arrowResultType
+      val actualResTpe = tpe2.arrowResultType
+      val subTypeResult = Type.isRegularSubType(expectedResTpe, actualResTpe, renv)
+
+      isMorePure(tpe1.arrowPurityType, tpe2.arrowPurityType, renv) &&
+        hasEffectSubset(tpe1.arrowEffectType, tpe2.arrowEffectType, renv) &&
+        superTypeArgs && subTypeResult
+
+    case (t1, t2) => t1 == t2
+
+  }
+
+  def isMorePure(pur1: Type, pur2: Type, renv: RigidityEnv)(implicit flix: Flix): Boolean = {
+    val loc = pur1.loc.asSynthetic
+    val pur3 = freshVar(Kind.Bool, loc)
+    val pur1pur3 = Type.mkAnd(pur1, pur3, loc)
+    Unification.unifiesWith(pur1pur3, pur2, renv)
+  }
+
+  def hasEffectSubset(eff1: Type, eff2: Type, renv: RigidityEnv)(implicit flix: Flix): Boolean = {
+    // The rule for effect sets is:
+    // S1 < S2 <==> exists S3 . S1 U S3 == S2
+    val loc = eff1.loc.asSynthetic
+    val eff3 = Type.freshVar(Kind.Effect, loc)
+    val eff1eff3 = Type.mkUnion(eff1, eff3, loc)
+    Unification.unifiesWith(eff1eff3, eff2, renv)
+  }
+
+  def isSubtypeOf(exp0: TypedAst.Expression, exp1: TypedAst.Expression, renv: RigidityEnv)(implicit flix: Flix): Boolean = {
```

Comment:
This one is very specific to the Safety phase, so it should stay there, and it can have a specific name like "canUpcast" with some explanation about which rules govern upcastability.